src/ast/desugar.rs

```diff
@@ -449,69 +587,63 @@ pub(crate) fn desugar_command(
         Command::Action(action) => flatten_actions(&vec![action], desugar)
             .into_iter()
             .map(NCommand::NormAction)
             .collect(),
-        Command::Run(config) => {
-            vec![NCommand::RunSchedule(NormSchedule::Run(
-                desugar_run_config(desugar, &config),
-            ))]
-        }
-        Command::Simplify { expr, config } => {
-            let fresh = desugar.get_fresh();
-            flatten_actions(&vec![Action::Let(fresh, expr)], desugar)
-                .into_iter()
-                .map(NCommand::NormAction)
-                .chain(
-                    vec![NCommand::Simplify {
-                        var: fresh,
-                        config: desugar_run_config(desugar, &config),
-                    }]
-                    .into_iter(),
-                )
-                .collect()
-        }
-        Command::Calc(idents, exprs) => desugar_calc(desugar, idents, exprs, seminaive),
+        Command::Simplify { expr, schedule } => desugar_simplify(desugar, &expr, &schedule),
+        Command::Calc(idents, exprs) => desugar_calc(desugar, idents, exprs, seminaive_transform)?,
         Command::RunSchedule(sched) => {
             vec![NCommand::RunSchedule(desugar_schedule(desugar, &sched))]
         }
-        Command::Extract { variants, e } => {
+        // TODO add variants to extract action
+        Command::Extract {
+            variants: _variants,
+            fact,
+        } => {
             let fresh = desugar.get_fresh();
-            flatten_actions(&vec![Action::Let(fresh, e)], desugar)
-                .into_iter()
-                .map(NCommand::NormAction)
-                .chain(
-                    vec![NCommand::Extract {
-                        variants,
-                        var: fresh,
-                    }]
-                    .into_iter(),
+            let fresh_ruleset = desugar.get_fresh();
+            let desugaring = if let Fact::Fact(Expr::Var(v)) = fact {
+                format!("(extract {v})")
+            } else {
+                format!(
+                    "(check {fact})
+                    (ruleset {fresh_ruleset})
+                    (rule ((= {fresh} {fact}))
+                          ((extract {fresh}))
+                          :ruleset {fresh_ruleset})
+                    (run {fresh_ruleset} 1)"
                 )
+            };
+
+            desugar
+                .desugar_program(
+                    desugar.parse_program(&desugaring).unwrap(),
+                    get_all_proofs,
+                    seminaive_transform,
+                )?
+                .into_iter()
+                .map(|cmd| cmd.command)
                 .collect()
         }
         Command::Check(facts) => {
-            let mut res = vec![NCommand::Check(flatten_facts(&facts, desugar))];
+            let res = vec![NCommand::Check(flatten_facts(&facts, desugar))];
 
             if get_all_proofs {
```

Comment:
There's lots of stuff commented out in this block. Should it be removed? If you want to leave it in, I think it'd help to add another comment as context.